src/wp-includes/wp-db.php

```diff
@@ -644,6 +653,17 @@ class wpdb {
 		'ANSI',
 	);
 
+	/**
+	 * Backwards compatibility, where wpdb::prepare() would not quote string placeholders with formatting.
+	 * They were used in the middle of longer strings, or as table name placeholders.
+	 * But they are risky, e.g. forgetting to add the quotes:
+	 *   $sql = $wpdb->prepare('WHERE (id = %s) OR (id = %1s) OR (id = %3$s)', ['id', 'id', 'id']);
+	 *
+	 * @since 6.0.0
+	 * @var bool
+	 */
+	public $unsafe_unquoted_parameters = true;
```

Comment:
This would be clearer with a more active name, such as `$allow_unsafe_unquoted_parameters`.

What's the reasoning behind making this true by default? That it'll eventually become false in a future release? Otherwise it seems unlikely ever to be false.

Reply:
Good point, I'm used to some projects and standards using "unsafe" as a prefix, but WordPress seems to prefer more human readable names.

It's `true` for backwards compatibility reasons ([note](https://github.com/WordPress/wordpress-develop/blob/26bb0d7479e1dd08dd982513ee6bee94f4c6f13d/src/wp-includes/wp-db.php#L1362)), I believe this was to allow the developer to do their own quoting, e.g. ``` `%1s` ``` for table/field names, or `"%1s something %1s"`. I intent to spend the next year or two checking and requesting updates to extensions (and core), so they avoid this risky feature (it's easy to get quoting wrong); where they can test by setting it to `false`, and when everyone is happy (well, as happy as they can be), we can switch the default to `false` (so anyone still using this feature can set it back to `true` with the knowledge that it's a problem), and hopefully one day (probably many years from now) we can remove it completely.

Reply:
Thanks for the explanation. That's about what I figured, good to have it stated clearly!

FYI the unsafe unquoted workaround is relatively uncommon, but not exactly rare. Some of these are false positive matches on js code: https://wpdirectory.net/search/01FY62PT1B54N9T5VTVDANX2H4

We could probably reach out to those developers in a systematic way. It'd be great to get their feedback on this solution. I'm curious to see how they use it in practice.

Reply:
>  where they can test by setting it to false, and when everyone is happy (well, as happy as they can be), we can switch the default to false (so anyone still using this feature can set it back to true with the knowledge that it's a problem), and hopefully one day (probably many years from now) we can remove it completely. 

FWIW, I tend to agree with @tellyworth here - It's unlikely to ever get flipped to `false` in it's current form, it's rare that we just drop support entirely unless forced to for security (which in this case, is IMHO unlikely).  If we remove it, developers should not just be able to flip the boolean back on because they want to.

If there isn't one already, a `_doing_it_wrong()`/deprecated notice triggered on these cases would not be out of the question, and it can eventually be removed if possible.

Reply:
I’d be open to adding a `_doing_it_wrong()`, and/or default to false, if everyone is happy to go straight to it. I would like to see this behaviour removed, but I don’t think we can break existing code without some kind of warning and fallback while extension developers update their risky code.

Reply:
> All placeholders MUST be left unquoted in the query string. A corresponding argument MUST be passed for each placeholder.
> 
> Note: There is one exception to the above: for compatibility with old behavior, numbered or formatted string placeholders (eg, %1$s, %5s) will not have quotes added by this function, so should be passed with appropriate quotes around them.

I don't think we can introduce the BC break this feature would cause without inadvertently opening up potential security issues, e.g.:
```
wp> $wpdb->allow_unsafe_unquoted_parameters = true;
=> bool(true)
wp> var_dump( $wpdb->prepare( "SELECT ... WHERE foo = '%-5s'", array( 'bar' ) ) );
string(30) "SELECT ... WHERE foo = 'bar  '"

wp> $wpdb->allow_unsafe_unquoted_parameters = false;
=> bool(false)
wp> $wpdb->prepare( "SELECT ... WHERE foo = '%-5s'", array( ' OR 1=1 -- -bar' ) );
=> string(42) "SELECT ... WHERE foo = '' OR 1=1 -- -bar''"
```

Additionally, I think numbered or formatted identifiers should also behave in the same manner as string placeholders and not be auto-quoted in backticks? I'm not a fan of their usage, but we should probably be consistent in how things are treated.

Reply:
That’s a good example of why we cannot simply set `allow_unsafe_unquoted_parameters` to false. Fortunately, most of the time it will result in an SQL parse error. The only working example I can think of is a normally numeric value, using string formatting, that can also contain evil SQL from the user. So while I’ll still work towards switching off this dangerous feature, I’ll need to raise awareness of the problem, get as many plugins corrected as possible, and work out the best transition at the time (based on feedback, numbers of remaining cases, probably using `_doing_it_wrong()` for a while, and we could even identify/block)… In the mean time, this patch allows security conscious people to switch this dangerous feature off and check their code.

As to adding this rarely used and dangerous feature to `%i` as well, while it would be “consistent”, it would only create more possibilities for mistakes… whereas the current patch means `%i` can only be used correctly (well, unless you want to create an SQL parse error).

Reply:
How about I make `$allow_unsafe_unquoted_parameters = true` private for now?

Then it cannot be accidentally be set to `false` (to ensure we're keeping the current behaviour, and not create any backwards compatibility problems); but it clearly calls this feature out as being unsafe, and gives me time to get plugins to update their code so they can either use `%i` or other (safer) approaches?

Then, after a couple of years, depending on how well I get on, with feedback from plugin authors, we can discuss how best to disable it.

Reply:
I've made `$allow_unsafe_unquoted_parameters = true` private for now, and added comments to explain its history, why the feature needs to be removed, and a note about why it cannot be simply switched to false. This avoids any backwards compatibility problems today, shows clear intent to remove, gives plugins time to use safer approaches (e.g. using `%i`), and gives us time to decide on the best removal process (i.e. in ~2 years time, we would see how well the plugin updates are going).

Reply:
As it's been a little while, and I think my changes have addressed everyones concerns, I'll resolve this conversation. Thank you.